lib/Parse/ParseExpr.cpp

```diff
@@ -2744,6 +2744,79 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
   Actions.ActOnBlockArguments(CaretLoc, DeclaratorInfo, getCurScope());
 }
 
+IdentifierInfo *Parser::getBoundsExpressionStart() {
+  if (Tok.getKind() == tok::identifier) {
+    IdentifierInfo *Ident = Tok.getIdentifierInfo();
+    if (Ident == Ident_bounds || Ident == Ident_byte_count || Ident == Ident_count) {
+      return Ident;
+    }
+  }
+  return nullptr;
+}
+
+ExprResult Parser::ParseBoundsExpression() {
+  IdentifierInfo *Ident = getBoundsExpressionStart();
+  if (!Ident) {
+    Diag(Tok, diag::err_expected_bounds_expr);
+    return ExprError();
```

Comment:
Is there a standard in clang for when you should provide more details with the expression error and when you shouldn't?  Below, if we hit a parser issue with the inner expressions (i.e. in ParseAssignmentExpression) will we preserve any inner error message or is our logic here stomping over that?